Dip/Dip/Dip.swift

```diff
@@ -42,8 +42,7 @@ public class DependencyContainer {
     case Int(IntegerLiteralType)
   }
   
-  private var dependencies = [DefinitionKey : Definition]()
-  private var lock: OSSpinLock = OS_SPINLOCK_INIT
```

Comment:
Yeah, seems safe to only lock on `resolve` as registrations are usually all done in one place at the start indeed.

But generally the philosophy in multi-threaded environments and when implementing thread-safety is to protect the access to a resource. When you add a lock to project a resource, you wrap _all_ the accesses to that resource with the lock, otherwise by principle the lock won't prevent _all_ concurrent access, so feels strange to not respect that usual pattern and only protect _some_ accesses to `dependencies` and not _not_… I'm hesitating.

(Anyway if we decide to either remove the lock or only lock resolutions and not registrations, that would be a breaking change that would need to be documented in the README and CHANGELOG)